drivers/spi/spi_smartbond.c

```diff
@@ -208,59 +619,435 @@ static int spi_smartbond_configure(const struct spi_smartbond_cfg *cfg,
+					return -EINVAL;
+				}
+				dma_start(config->rx_dma_ctrl, config->rx_dma_chan);
+				dma_start(config->tx_dma_ctrl, config->tx_dma_chan);
+
+				k_sem_take(&data->tx_dma_sync, K_FOREVER);
+				k_sem_take(&data->rx_dma_sync, K_FOREVER);
+			}
+			break;
+		case SPI_SMARTBOND_TRANSFER_NONE:
+			__fallthrough;
+		default:
+			__ASSERT_MSG_INFO("Invalid transfer mode");
+			break;
+		}
+
+		spi_smartbond_transfer_mode_check_and_update(dev);
+	} while (data->transfer_mode != SPI_SMARTBOND_TRANSFER_NONE);
+
+	return 0;
+}
+#endif
+
 static int spi_smartbond_transceive(const struct device *dev, const struct spi_config *spi_cfg,
 				    const struct spi_buf_set *tx_bufs,
 				    const struct spi_buf_set *rx_bufs)
 {
 	const struct spi_smartbond_cfg *cfg = dev->config;
 	struct spi_smartbond_data *data = dev->data;
 	struct spi_context *ctx = &data->ctx;
-	uint32_t bitmask;
 	int rc;
 
 	spi_smartbond_pm_policy_state_lock_get(data);
 
 	spi_context_lock(&data->ctx, false, NULL, NULL, spi_cfg);
+
 	rc = spi_smartbond_configure(cfg, data, spi_cfg);
 	if (rc == 0) {
 		spi_context_buffers_setup(ctx, tx_bufs, rx_bufs, data->dfs);
 		spi_context_cs_control(ctx, true);
 
-		bitmask = ~((~0UL) << SPI_WORD_SIZE_GET(data->ctx.config->operation));
 		while (spi_context_tx_buf_on(ctx) || spi_context_rx_buf_on(ctx)) {
 			if (spi_context_tx_buf_on(ctx)) {
-				cfg->regs->SPI_RX_TX_REG = (*(uint32_t *)ctx->tx_buf) & bitmask;
+				spi_smartbond_write_word(dev);
 				spi_context_update_tx(ctx, data->dfs, 1);
 			} else {
-				cfg->regs->SPI_RX_TX_REG = 0UL;
+				spi_smartbond_write_dummy(dev);
 			}
 
 			while (!(cfg->regs->SPI_CTRL_REG & SPI_SPI_CTRL_REG_SPI_INT_BIT_Msk)) {
 			};
 			if (spi_context_rx_buf_on(ctx)) {
-				(*(uint32_t *)ctx->rx_buf) = cfg->regs->SPI_RX_TX_REG & bitmask;
+				spi_smartbond_read_word(dev);
 				spi_context_update_rx(ctx, data->dfs, 1);
 			} else {
-				(void)cfg->regs->SPI_RX_TX_REG;
+				spi_smartbond_read_discard(dev);
 			}
 			cfg->regs->SPI_CLEAR_INT_REG = 1UL;
 		}
+
+		spi_context_cs_control(ctx, false);
 	}
-	spi_context_cs_control(ctx, false);
 	spi_context_release(&data->ctx, rc);
 
 	spi_smartbond_pm_policy_state_lock_put(data);
 
 	return rc;
 }
+
 #ifdef CONFIG_SPI_ASYNC
 static int spi_smartbond_transceive_async(const struct device *dev,
 					  const struct spi_config *spi_cfg,
 					  const struct spi_buf_set *tx_bufs,
 					  const struct spi_buf_set *rx_bufs, spi_callback_t cb,
 					  void *userdata)
 {
-	return -ENOTSUP;
+	const struct spi_smartbond_cfg *cfg = dev->config;
+	struct spi_smartbond_data *data = dev->data;
+	struct spi_context *ctx = &data->ctx;
+	int rc;
+
+	spi_context_lock(ctx, true, cb, userdata, spi_cfg);
+
+	rc = spi_smartbond_configure(cfg, data, spi_cfg);
+	if (rc == 0) {
+		spi_smartbond_pm_policy_state_lock_get(data);
+		spi_context_buffers_setup(ctx, tx_bufs, rx_bufs, data->dfs);
+		spi_context_cs_control(ctx, true);
+
+#ifndef CONFIG_SPI_SMARTBOND_DMA
```

Comment:
Are you sure you want to, at built time, enable DMA or interrupt mode as exclusive mode?

What if a controller has a DMA configuration in DTS, but another one has not?

Check how spi_ll_stm32.c does this.

Reply:
> What if a controller has a DMA configuration in DTS, but another one has not?

Fixed, `COND_CODE_1` is used to check if  `dmas` cells are defined in DTS. 

> Are you sure you want to, at built time, enable DMA or interrupt mode as exclusive mode?

I get the point. Since the DMA approach is actually a blocking operation, it should be bound to the blocking API. At the same time non-blocking APIs are available using the interrupt-driven approach. Will update it and let you know once this is done.

Reply:
Fixed

Reply:
it's not a blocking/non-blocking issue (DMA is interrupt based also), I was not clear enough describing the issue I see. It's because you may have 2 SPI controllers defined in DTS, one with DMA and the other without. The issue with your first approach was that you excluded this possibility as soon as the config option for DMA is enabled.

SPI_ASYNC it not tightened to interrupt mode actually, it's enabling part of the SPI API enabling the user to not block on the transceive call.
You can have a blocking transceive call, but still running in interrupt mode. Actually and preferably, all calls should be running under interrupt mode.

Reply:
Thanks for the thorough analysis provided. It's now clear to me.